sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala

```diff
@@ -0,0 +1,249 @@
+    val values = (0 until genericArrayData.numElements()).map(genericArrayData.get(_, StringType).
+      asInstanceOf[UTF8String]).toVector
+    val nGrams = getNGrams(values, n)
+    nGrams.foreach(buffer.add(_))
+    buffer.trim()
+  }
+  override def update(buffer: NGramBuffer, inputRow: InternalRow): NGramBuffer = {
+    if (isArrayOfString) {
+      updateArray(child.eval(inputRow).asInstanceOf[GenericArrayData], buffer, inputRow)
+    } else {
+      val arrayOfArray = child.eval(inputRow).asInstanceOf[GenericArrayData]
+      for (i <- 0 until arrayOfArray.numElements()) {
+        updateArray(arrayOfArray.getArray(i).asInstanceOf[GenericArrayData], buffer, inputRow)
+      }
+    }
+    buffer
+  }
+
+  override def merge(buffer: NGramBuffer, input: NGramBuffer): NGramBuffer = {
+    buffer.merge(input)
+    buffer.trim()
+    buffer
+  }
+
+  override def eval(buffer: NGramBuffer): Any = {
+    val topKNGrams = buffer.getTopKNGrams().map((keyValuePair: (Vector[UTF8String], Double)) => {
+      val arrayKey = new GenericArrayData(Vector(new GenericArrayData(keyValuePair._1)))
+      val arrayValue = new GenericArrayData(Vector(keyValuePair._2))
+      new ArrayBasedMapData(arrayKey, arrayValue)
+    }).toVector
+    new GenericArrayData(topKNGrams)
+  }
+
+  private def getNGrams(values: Vector[UTF8String], n: Int): Vector[Vector[UTF8String]] = {
+    if (values.length >= n) {
+      values.sliding(n).toVector
+    } else {
+      Vector()
+    }
+  }
+
+  override def withNewMutableAggBufferOffset(newOffset: Int): NGrams =
+    copy(mutableAggBufferOffset = newOffset)
+
+  override def withNewInputAggBufferOffset(newOffset: Int): NGrams =
+    copy(inputAggBufferOffset = newOffset)
+
+  override def children: Seq[Expression] = Seq(child, nExpression, kExpression, accuracyExpression)
+
+  // Returns null for empty inputs
+  override def nullable: Boolean = true
+
+  override def dataType: DataType = ArrayType(MapType(ArrayType(StringType), DoubleType))
+
+  override def prettyName: String = "ngrams"
+
+  override def serialize(obj: NGramBuffer): Array[Byte] = {
+    NGrams.serializer.serialize(obj)
+  }
+
+  override def deserialize(bytes: Array[Byte]): NGramBuffer = {
+    NGrams.serializer.deserialize(bytes)
+  }
+}
+
+object NGrams {
+
+  val DEFAULT_ACCURACY: Int = 1000
+
+  private def getAccuracy(kExpression: Expression, accuracyExpression: Expression): Expression = {
+    val accuracy = accuracyExpression.eval().asInstanceOf[Int]
+    val k = kExpression.eval().asInstanceOf[Int]
+    Literal(accuracy.max(DEFAULT_ACCURACY / k))
+  }
+
+  val kryoSerializer: KryoSerializer = new KryoSerializer(new SparkConf())
+
+  import collection.JavaConverters._
+
+  class NGramBuffer(val n: Int,
+                    val k: Int,
+                    val precisionFactor: Int,
+                    val frequencyMap: HashMap[Vector[UTF8String], Double]) {
```

Comment:
You are right, it doesn't need to be. By the way, what's the community's major concerns about merging the [expression](https://github.com/apache/spark/pull/16476)/aggregate expression PRs? The performance?